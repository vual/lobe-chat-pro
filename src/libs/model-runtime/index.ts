export { LobeAnthropicAI } from './anthropic';
export { LobeAzureAI } from './azureai';
export { LobeAzureOpenAI } from './azureOpenai';
export * from './BaseAI';
export { LobeBedrockAI } from './bedrock';
export { LobeDeepSeekAI } from './deepseek';
export * from './error';
export { LobeGoogleAI } from './google';
export { LobeGroq } from './groq';
export * from './helpers';
export { LobeMinimaxAI } from './minimax';
export { LobeMistralAI } from './mistral';
export { default as AgentRuntime } from './ModelRuntime';
export { LobeMoonshotAI } from './moonshot';
export { LobeOllamaAI } from './ollama';
export { LobeOpenAI } from './openai';
export { LobeOpenRouterAI } from './openrouter';
export { LobePerplexityAI } from './perplexity';
export { LobeQwenAI } from './qwen';
export { LobeTogetherAI } from './togetherai';
export * from './types';
export { AgentRuntimeError } from './utils/createError';
export { createOpenAICompatibleRuntime } from './utils/openaiCompatibleFactory';
export { pruneReasoningPayload } from './utils/openaiHelpers';
export { LobeVolcengineAI } from './volcengine';
export { LobeZeroOneAI } from './zeroone';
export { LobeZhipuAI } from './zhipu';
