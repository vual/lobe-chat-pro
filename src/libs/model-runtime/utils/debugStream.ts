// no need to introduce a package to get the current time as this module is just a debug utility
const getTime = () => {
  const date = new Date();
  return `${date.getFullYear()}-${date.getMonth() + 1}-${date.getDate()} ${date.getHours()}:${date.getMinutes()}:${date.getSeconds()}.${date.getMilliseconds()}`;
};

export const debugStream = async (stream: ReadableStream) => {
  let finished = false;
  let chunk = 0;
  let chunkValue: any;
  const decoder = new TextDecoder();

  const reader = stream.getReader();

  console.log(`[stream start] ${getTime()}`);

  while (!finished) {
    try {
      const { value, done } = await reader.read();

      if (done) {
        console.log(`[stream finished] total chunks: ${chunk}\n`);
        finished = true;
        break;
      }

      chunkValue = value;

      // if the value is ArrayBuffer, we need to decode it
      if ('byteLength' in value) {
        chunkValue = decoder.decode(value, { stream: true });
      } else if (typeof value !== 'string') {
        chunkValue = JSON.stringify(value);
      }

      console.log(`[chunk ${chunk}] ${getTime()}`);
      console.log(chunkValue);
      console.log('');

      finished = done;
      chunk++;
    } catch (e) {
      finished = true;
      console.error('[debugStream error]', e);
      console.error('[error chunk value:]', chunkValue);
    }
  }
};

export const debugResponse = (response: any) => {
  console.log(`\n[no stream response] ${getTime()}\n`);
  console.log(JSON.stringify(response) + '\n');
};
