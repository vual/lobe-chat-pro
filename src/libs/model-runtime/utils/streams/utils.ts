export const createReadableStream = <T>(chunks: T[]) =>
  new ReadableStream({
    start(controller) {
      chunks.forEach((chunk) => controller.enqueue(chunk));

      controller.close();
    },
  });

export const readStreamChunk = async (stream: ReadableStream) => {
  const decoder = new TextDecoder();
  const chunks = [];

  // @ts-ignore
  for await (const chunk of stream) {
    chunks.push(decoder.decode(chunk, { stream: true }));
  }

  return chunks;
};
