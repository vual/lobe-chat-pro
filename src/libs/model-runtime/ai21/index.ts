import { ModelProvider } from '../types';
import { createOpenAICompatibleRuntime } from '../utils/openaiCompatibleFactory';

export const LobeAi21AI = createOpenAICompatibleRuntime({
  baseURL: 'https://api.ai21.com/studio/v1',
  chatCompletion: {
    handlePayload: (payload) => {
      return {
        ...payload,
        stream: !payload.tools,
      } as any;
    },
  },
  debug: {
    chatCompletion: () => process.env.DEBUG_AI21_CHAT_COMPLETION === '1',
  },
  provider: ModelProvider.Ai21,
});
