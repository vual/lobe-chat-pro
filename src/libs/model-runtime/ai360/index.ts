import type { ChatModelCard } from '@/types/llm';

import { ModelProvider } from '../types';
import { createOpenAICompatibleRuntime } from '../utils/openaiCompatibleFactory';

export interface Ai360ModelCard {
  id: string;
  max_tokens: number;
  total_tokens: number;
}

export const LobeAi360AI = createOpenAICompatibleRuntime({
  baseURL: 'https://api.360.cn/v1',
  chatCompletion: {
    handlePayload: (payload) => {
      const { enabledSearch, tools, ...rest } = payload;

      const ai360Tools = enabledSearch
        ? [
            ...(tools || []),
            {
              type: 'web_search',
              web_search: {
                search_mode: 'auto',
              },
            },
          ]
        : tools;

      return {
        ...rest,
        stream: !ai360Tools,
        tools: ai360Tools,
      } as any;
    },
  },
  debug: {
    chatCompletion: () => process.env.DEBUG_AI360_CHAT_COMPLETION === '1',
  },
  models: async ({ client }) => {
    const { LOBE_DEFAULT_MODEL_LIST } = await import('@/config/aiModels');

    const reasoningKeywords = ['360gpt2-o1', '360zhinao2-o1'];

    const modelsPage = (await client.models.list()) as any;
    const modelList: Ai360ModelCard[] = modelsPage.data;

    return modelList
      .map((model) => {
        const knownModel = LOBE_DEFAULT_MODEL_LIST.find(
          (m) => model.id.toLowerCase() === m.id.toLowerCase(),
        );

        return {
          contextWindowTokens: model.total_tokens,
          displayName: knownModel?.displayName ?? undefined,
          enabled: knownModel?.enabled || false,
          functionCall: model.id === '360gpt-pro' || knownModel?.abilities?.functionCall || false,
          id: model.id,
          maxTokens: typeof model.max_tokens === 'number' ? model.max_tokens : undefined,
          reasoning:
            reasoningKeywords.some((keyword) => model.id.toLowerCase().includes(keyword)) ||
            knownModel?.abilities?.reasoning ||
            false,
          vision: knownModel?.abilities?.vision || false,
        };
      })
      .filter(Boolean) as ChatModelCard[];
  },
  provider: ModelProvider.Ai360,
});
