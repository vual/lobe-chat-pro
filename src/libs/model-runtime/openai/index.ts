import { responsesAPIModels } from '@/const/models';

import { ChatStreamPayload, ModelProvider } from '../types';
import { processMultiProviderModelList } from '../utils/modelParse';
import { createOpenAICompatibleRuntime } from '../utils/openaiCompatibleFactory';
import { pruneReasoningPayload } from '../utils/openaiHelpers';

export interface OpenAIModelCard {
  id: string;
}

const prunePrefixes = ['o1', 'o3', 'o4', 'codex', 'computer-use'];

const oaiSearchContextSize = process.env.OPENAI_SEARCH_CONTEXT_SIZE; // low, medium, high

export const LobeOpenAI = createOpenAICompatibleRuntime({
  baseURL: 'https://api.openai.com/v1',
  chatCompletion: {
    handlePayload: (payload) => {
      const { enabledSearch, model, ...rest } = payload;

      if (responsesAPIModels.has(model) || enabledSearch) {
        return { ...rest, apiMode: 'responses', enabledSearch, model } as ChatStreamPayload;
      }

      if (prunePrefixes.some((prefix) => model.startsWith(prefix))) {
        return pruneReasoningPayload(payload) as any;
      }

      if (model.includes('-search-')) {
        return {
          ...rest,
          frequency_penalty: undefined,
          model,
          presence_penalty: undefined,
          stream: payload.stream ?? true,
          temperature: undefined,
          top_p: undefined,
          ...(oaiSearchContextSize && {
            web_search_options: {
              search_context_size: oaiSearchContextSize,
            },
          }),
        } as any;
      }

      return { ...rest, model, stream: payload.stream ?? true };
    },
  },
  debug: {
    chatCompletion: () => process.env.DEBUG_OPENAI_CHAT_COMPLETION === '1',
    responses: () => process.env.DEBUG_OPENAI_RESPONSES === '1',
  },
  models: async ({ client }) => {
    const modelsPage = (await client.models.list()) as any;
    const modelList: OpenAIModelCard[] = modelsPage.data;

    // 自动检测模型提供商并选择相应配置
    return processMultiProviderModelList(modelList);
  },
  provider: ModelProvider.OpenAI,
  responses: {
    handlePayload: (payload) => {
      const { enabledSearch, model, tools, ...rest } = payload;

      const openaiTools = enabledSearch
        ? [
            ...(tools || []),
            {
              type: 'web_search_preview',
              ...(oaiSearchContextSize && {
                search_context_size: oaiSearchContextSize,
              }),
            },
          ]
        : tools;

      if (prunePrefixes.some((prefix) => model.startsWith(prefix))) {
        if (!payload.reasoning) {
          payload.reasoning = { summary: 'auto' };
        } else {
          payload.reasoning.summary = 'auto';
        }

        // computer-use series must set truncation as auto
        if (model.startsWith('computer-use')) {
          payload.truncation = 'auto';
        }

        return pruneReasoningPayload(payload) as any;
      }

      return { ...rest, model, stream: payload.stream ?? true, tools: openaiTools } as any;
    },
  },
});
