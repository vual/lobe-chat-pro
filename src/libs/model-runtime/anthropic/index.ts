import Anthropic, { ClientOptions } from '@anthropic-ai/sdk';

import type { ChatModelCard } from '@/types/llm';

import { LobeRuntimeAI } from '../BaseAI';
import { AgentRuntimeErrorType } from '../error';
import {
  type ChatCompletionErrorPayload,
  ChatMethodOptions,
  ChatStreamPayload,
  ModelProvider,
} from '../types';
import { buildAnthropicMessages, buildAnthropicTools } from '../utils/anthropicHelpers';
import { AgentRuntimeError } from '../utils/createError';
import { debugStream } from '../utils/debugStream';
import { desensitizeUrl } from '../utils/desensitizeUrl';
import { StreamingResponse } from '../utils/response';
import { AnthropicStream } from '../utils/streams';
import { handleAnthropicError } from './handleAnthropicError';

export interface AnthropicModelCard {
  display_name: string;
  id: string;
}

type anthropicTools = Anthropic.Tool | Anthropic.WebSearchTool20250305;

const modelsWithSmallContextWindow = new Set(['claude-3-opus-20240229', 'claude-3-haiku-20240307']);

const DEFAULT_BASE_URL = 'https://api.anthropic.com';

interface AnthropicAIParams extends ClientOptions {
  id?: string;
}

export class LobeAnthropicAI implements LobeRuntimeAI {
  private client: Anthropic;

  baseURL: string;
  apiKey?: string;
  private id: string;

  private isDebug() {
    return process.env.DEBUG_ANTHROPIC_CHAT_COMPLETION === '1';
  }

  constructor({ apiKey, baseURL = DEFAULT_BASE_URL, id, ...res }: AnthropicAIParams = {}) {
    if (!apiKey) throw AgentRuntimeError.createError(AgentRuntimeErrorType.InvalidProviderAPIKey);

    const betaHeaders = process.env.ANTHROPIC_BETA_HEADERS;

    this.client = new Anthropic({
      apiKey,
      baseURL,
      ...(betaHeaders ? { defaultHeaders: { 'anthropic-beta': betaHeaders } } : {}),
      ...res,
    });
    this.baseURL = this.client.baseURL;
    this.apiKey = apiKey;
    this.id = id || ModelProvider.Anthropic;
  }

  async chat(payload: ChatStreamPayload, options?: ChatMethodOptions) {
    try {
      const anthropicPayload = await this.buildAnthropicPayload(payload);
      const inputStartAt = Date.now();

      if (this.isDebug()) {
        console.log('[requestPayload]');
        console.log(JSON.stringify(anthropicPayload), '\n');
      }

      const response = await this.client.messages.create(
        {
          ...anthropicPayload,
          metadata: options?.user ? { user_id: options?.user } : undefined,
          stream: true,
        },
        {
          signal: options?.signal,
        },
      );

      const [prod, debug] = response.tee();

      if (this.isDebug()) {
        debugStream(debug.toReadableStream()).catch(console.error);
      }

      return StreamingResponse(
        AnthropicStream(prod, { callbacks: options?.callback, inputStartAt }),
        {
          headers: options?.headers,
        },
      );
    } catch (error) {
      throw this.handleError(error);
    }
  }

  private async buildAnthropicPayload(payload: ChatStreamPayload) {
    const {
      messages,
      model,
      max_tokens,
      temperature,
      top_p,
      tools,
      thinking,
      enabledContextCaching = true,
      enabledSearch,
    } = payload;

    const { default: anthropicModels } = await import('@/config/aiModels/anthropic');
    const modelConfig = anthropicModels.find((m) => m.id === model);
    const defaultMaxOutput = modelConfig?.maxOutput;

    // 配置优先级：用户设置 > 模型配置 > 硬编码默认值
    const getMaxTokens = () => {
      if (max_tokens) return max_tokens;
      if (defaultMaxOutput) return defaultMaxOutput;
      return undefined;
    };

    const system_message = messages.find((m) => m.role === 'system');
    const user_messages = messages.filter((m) => m.role !== 'system');

    const systemPrompts = !!system_message?.content
      ? ([
          {
            cache_control: enabledContextCaching ? { type: 'ephemeral' } : undefined,
            text: system_message?.content as string,
            type: 'text',
          },
        ] as Anthropic.TextBlockParam[])
      : undefined;

    const postMessages = await buildAnthropicMessages(user_messages, { enabledContextCaching });

    let postTools: anthropicTools[] | undefined = buildAnthropicTools(tools, {
      enabledContextCaching,
    });

    if (enabledSearch) {
      // Limit the number of searches per request
      const maxUses = process.env.ANTHROPIC_MAX_USES;

      const webSearchTool: Anthropic.WebSearchTool20250305 = {
        name: 'web_search',
        type: 'web_search_20250305',
        ...(maxUses &&
          Number.isInteger(Number(maxUses)) &&
          Number(maxUses) > 0 && {
            max_uses: Number(maxUses),
          }),
      };

      // 如果已有工具，则添加到现有工具列表中；否则创建新的工具列表
      if (postTools && postTools.length > 0) {
        postTools = [...postTools, webSearchTool];
      } else {
        postTools = [webSearchTool];
      }
    }

    if (!!thinking && thinking.type === 'enabled') {
      const maxTokens = getMaxTokens() || 32_000; // Claude Opus 4 has minimum maxOutput

      // `temperature` may only be set to 1 when thinking is enabled.
      // `top_p` must be unset when thinking is enabled.
      return {
        max_tokens: maxTokens,
        messages: postMessages,
        model,
        system: systemPrompts,
        thinking: {
          ...thinking,
          budget_tokens: thinking?.budget_tokens
            ? Math.min(thinking.budget_tokens, maxTokens - 1) // `max_tokens` must be greater than `thinking.budget_tokens`.
            : 1024,
        },
        tools: postTools,
      } satisfies Anthropic.MessageCreateParams;
    }

    return {
      // claude 3 series model hax max output token of 4096, 3.x series has 8192
      // https://docs.anthropic.com/en/docs/about-claude/models/all-models#:~:text=200K-,Max%20output,-Normal%3A
      max_tokens: getMaxTokens() || (modelsWithSmallContextWindow.has(model) ? 4096 : 8192),
      messages: postMessages,
      model,
      system: systemPrompts,
      temperature: payload.temperature !== undefined ? temperature / 2 : undefined,
      tools: postTools,
      top_p,
    } satisfies Anthropic.MessageCreateParams;
  }

  async models() {
    const { LOBE_DEFAULT_MODEL_LIST } = await import('@/config/aiModels');

    const url = `${this.baseURL}/v1/models`;
    const response = await fetch(url, {
      headers: {
        'anthropic-version': '2023-06-01',
        'x-api-key': `${this.apiKey}`,
      },
      method: 'GET',
    });
    const json = await response.json();

    const modelList: AnthropicModelCard[] = json['data'];

    return modelList
      .map((model) => {
        const knownModel = LOBE_DEFAULT_MODEL_LIST.find(
          (m) => model.id.toLowerCase() === m.id.toLowerCase(),
        );

        return {
          contextWindowTokens: knownModel?.contextWindowTokens ?? undefined,
          displayName: model.display_name,
          enabled: knownModel?.enabled || false,
          functionCall:
            model.id.toLowerCase().includes('claude-3') ||
            knownModel?.abilities?.functionCall ||
            false,
          id: model.id,
          reasoning: knownModel?.abilities?.reasoning || false,
          vision:
            (model.id.toLowerCase().includes('claude-3') &&
              !model.id.toLowerCase().includes('claude-3-5-haiku')) ||
            knownModel?.abilities?.vision ||
            false,
        };
      })
      .filter(Boolean) as ChatModelCard[];
  }

  private handleError(error: any): ChatCompletionErrorPayload {
    let desensitizedEndpoint = this.baseURL;

    if (this.baseURL !== DEFAULT_BASE_URL) {
      desensitizedEndpoint = desensitizeUrl(this.baseURL);
    }

    if ('status' in (error as any)) {
      switch ((error as Response).status) {
        case 401: {
          throw AgentRuntimeError.chat({
            endpoint: desensitizedEndpoint,
            error: error as any,
            errorType: AgentRuntimeErrorType.InvalidProviderAPIKey,
            provider: this.id,
          });
        }

        case 403: {
          throw AgentRuntimeError.chat({
            endpoint: desensitizedEndpoint,
            error: error as any,
            errorType: AgentRuntimeErrorType.LocationNotSupportError,
            provider: this.id,
          });
        }
        default: {
          break;
        }
      }
    }

    const { errorResult } = handleAnthropicError(error);

    throw AgentRuntimeError.chat({
      endpoint: desensitizedEndpoint,
      error: errorResult,
      errorType: AgentRuntimeErrorType.ProviderBizError,
      provider: this.id,
    });
  }
}

export default LobeAnthropicAI;
