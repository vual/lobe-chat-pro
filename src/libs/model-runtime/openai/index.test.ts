// @vitest-environment node
import OpenAI from 'openai';
import { Mock, afterEach, beforeEach, describe, expect, it, vi } from 'vitest';

// 引入模块以便于对函数进行spy
import { ChatStreamCallbacks, LobeOpenAICompatibleRuntime } from '@/libs/model-runtime';

import * as debugStreamModule from '../utils/debugStream';
import officalOpenAIModels from './fixtures/openai-models.json';
import { LobeOpenAI } from './index';

// Mock the console.error to avoid polluting test output
vi.spyOn(console, 'error').mockImplementation(() => {});

describe('LobeOpenAI', () => {
  let instance: LobeOpenAICompatibleRuntime;

  beforeEach(() => {
    instance = new LobeOpenAI({ apiKey: 'test' });

    // 使用 vi.spyOn 来模拟 chat.completions.create 方法
    vi.spyOn(instance['client'].chat.completions, 'create').mockResolvedValue(
      new ReadableStream() as any,
    );
    vi.spyOn(instance['client'].models, 'list').mockResolvedValue({ data: [] } as any);
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  describe('chat', () => {
    it('should return a StreamingTextResponse on successful API call', async () => {
      // Arrange
      const mockStream = new ReadableStream();
      const mockResponse = Promise.resolve(mockStream);

      (instance['client'].chat.completions.create as Mock).mockResolvedValue(mockResponse);

      // Act
      const result = await instance.chat({
        messages: [{ content: 'Hello', role: 'user' }],
        model: 'text-davinci-003',
        temperature: 0,
      });

      // Assert
      expect(result).toBeInstanceOf(Response);
    });

    describe('Error', () => {
      it('should return ProviderBizError with an openai error response when OpenAI.APIError is thrown', async () => {
        // Arrange
        const apiError = new OpenAI.APIError(
          400,
          {
            status: 400,
            error: {
              message: 'Bad Request',
            },
          },
          'Error message',
          {},
        );

        vi.spyOn(instance['client'].chat.completions, 'create').mockRejectedValue(apiError);

        // Act
        try {
          await instance.chat({
            messages: [{ content: 'Hello', role: 'user' }],
            model: 'text-davinci-003',
            temperature: 0,
          });
        } catch (e) {
          expect(e).toEqual({
            endpoint: 'https://api.openai.com/v1',
            error: {
              error: { message: 'Bad Request' },
              status: 400,
            },
            errorType: 'ProviderBizError',
            provider: 'openai',
          });
        }
      });

      it('should throw AgentRuntimeError with NoOpenAIAPIKey if no apiKey is provided', async () => {
        try {
          new LobeOpenAI({});
        } catch (e) {
          expect(e).toEqual({ errorType: 'InvalidProviderAPIKey' });
        }
      });

      it('should return ProviderBizError with the cause when OpenAI.APIError is thrown with cause', async () => {
        // Arrange
        const errorInfo = {
          stack: 'abc',
          cause: {
            message: 'api is undefined',
          },
        };
        const apiError = new OpenAI.APIError(400, errorInfo, 'module error', {});

        vi.spyOn(instance['client'].chat.completions, 'create').mockRejectedValue(apiError);

        // Act
        try {
          await instance.chat({
            messages: [{ content: 'Hello', role: 'user' }],
            model: 'text-davinci-003',
            temperature: 0,
          });
        } catch (e) {
          expect(e).toEqual({
            endpoint: 'https://api.openai.com/v1',
            error: {
              cause: { message: 'api is undefined' },
              stack: 'abc',
            },
            errorType: 'ProviderBizError',
            provider: 'openai',
          });
        }
      });

      it('should return ProviderBizError with an cause response with desensitize Url', async () => {
        // Arrange
        const errorInfo = {
          stack: 'abc',
          cause: { message: 'api is undefined' },
        };
        const apiError = new OpenAI.APIError(400, errorInfo, 'module error', {});

        instance = new LobeOpenAI({
          apiKey: 'test',

          baseURL: 'https://api.abc.com/v1',
        });

        vi.spyOn(instance['client'].chat.completions, 'create').mockRejectedValue(apiError);

        // Act
        try {
          await instance.chat({
            messages: [{ content: 'Hello', role: 'user' }],
            model: 'gpt-3.5-turbo',
            temperature: 0,
          });
        } catch (e) {
          expect(e).toEqual({
            endpoint: 'https://api.***.com/v1',
            error: {
              cause: { message: 'api is undefined' },
              stack: 'abc',
            },
            errorType: 'ProviderBizError',
            provider: 'openai',
          });
        }
      });

      it('should return AgentRuntimeError for non-OpenAI errors', async () => {
        // Arrange
        const genericError = new Error('Generic Error');

        vi.spyOn(instance['client'].chat.completions, 'create').mockRejectedValue(genericError);

        // Act
        try {
          await instance.chat({
            messages: [{ content: 'Hello', role: 'user' }],
            model: 'text-davinci-003',
            temperature: 0,
          });
        } catch (e) {
          expect(e).toEqual({
            endpoint: 'https://api.openai.com/v1',
            errorType: 'AgentRuntimeError',
            provider: 'openai',
            error: {
              name: genericError.name,
              cause: genericError.cause,
              message: genericError.message,
              stack: genericError.stack,
            },
          });
        }
      });
    });

    describe('DEBUG', () => {
      it('should call debugStream and return StreamingTextResponse when DEBUG_OPENAI_CHAT_COMPLETION is 1', async () => {
        // Arrange
        const mockProdStream = new ReadableStream() as any; // 模拟的 prod 流
        const mockDebugStream = new ReadableStream({
          start(controller) {
            controller.enqueue('Debug stream content');
            controller.close();
          },
        }) as any;
        mockDebugStream.toReadableStream = () => mockDebugStream; // 添加 toReadableStream 方法

        // 模拟 chat.completions.create 返回值，包括模拟的 tee 方法
        (instance['client'].chat.completions.create as Mock).mockResolvedValue({
          tee: () => [mockProdStream, { toReadableStream: () => mockDebugStream }],
        });

        // 保存原始环境变量值
        const originalDebugValue = process.env.DEBUG_OPENAI_CHAT_COMPLETION;

        // 模拟环境变量
        process.env.DEBUG_OPENAI_CHAT_COMPLETION = '1';
        vi.spyOn(debugStreamModule, 'debugStream').mockImplementation(() => Promise.resolve());

        // 执行测试
        // 运行你的测试函数，确保它会在条件满足时调用 debugStream
        // 假设的测试函数调用，你可能需要根据实际情况调整
        await instance.chat({
          messages: [{ content: 'Hello', role: 'user' }],
          model: 'text-davinci-003',
          temperature: 0,
        });

        // 验证 debugStream 被调用
        expect(debugStreamModule.debugStream).toHaveBeenCalled();

        // 恢复原始环境变量值
        process.env.DEBUG_OPENAI_CHAT_COMPLETION = originalDebugValue;
      });
    });
  });

  describe('models', () => {
    it('should get models', async () => {
      // mock the models.list method
      (instance['client'].models.list as Mock).mockResolvedValue({ data: officalOpenAIModels });

      const list = await instance.models();

      expect(list).toMatchSnapshot();
    });
  });
});
