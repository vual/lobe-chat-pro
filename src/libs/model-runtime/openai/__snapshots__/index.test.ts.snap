// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LobeOpenAI > models > should get models 1`] = `
[
  {
    "contextWindowTokens": undefined,
    "displayName": "Whisper",
    "enabled": false,
    "functionCall": false,
    "id": "whisper-1",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": undefined,
    "displayName": "davinci-002",
    "enabled": false,
    "functionCall": false,
    "id": "davinci-002",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": 16385,
    "displayName": "GPT-3.5 Turbo",
    "enabled": false,
    "functionCall": true,
    "id": "gpt-3.5-turbo",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": undefined,
    "displayName": "DALL·E 2",
    "enabled": false,
    "functionCall": false,
    "id": "dall-e-2",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": undefined,
    "displayName": "gpt-3.5-turbo-16k",
    "enabled": false,
    "functionCall": false,
    "id": "gpt-3.5-turbo-16k",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": undefined,
    "displayName": "tts-1-hd-1106",
    "enabled": false,
    "functionCall": false,
    "id": "tts-1-hd-1106",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": undefined,
    "displayName": "TTS-1 HD",
    "enabled": false,
    "functionCall": false,
    "id": "tts-1-hd",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": undefined,
    "displayName": "gpt-3.5-turbo-16k-0613",
    "enabled": false,
    "functionCall": false,
    "id": "gpt-3.5-turbo-16k-0613",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": 8192,
    "displayName": "Text Embedding 3 Large",
    "enabled": false,
    "functionCall": false,
    "id": "text-embedding-3-large",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": undefined,
    "displayName": "gpt-4-1106-vision-preview",
    "enabled": false,
    "functionCall": false,
    "id": "gpt-4-1106-vision-preview",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": undefined,
    "displayName": "gpt-3.5-turbo-instruct-0914",
    "enabled": false,
    "functionCall": false,
    "id": "gpt-3.5-turbo-instruct-0914",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": 128000,
    "displayName": "GPT-4 Turbo Preview 0125",
    "enabled": false,
    "functionCall": true,
    "id": "gpt-4-0125-preview",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": 128000,
    "displayName": "GPT-4 Turbo Preview",
    "enabled": false,
    "functionCall": true,
    "id": "gpt-4-turbo-preview",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": 4096,
    "displayName": "GPT-3.5 Turbo Instruct",
    "enabled": false,
    "functionCall": false,
    "id": "gpt-3.5-turbo-instruct",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": undefined,
    "displayName": "gpt-3.5-turbo-0301",
    "enabled": false,
    "functionCall": false,
    "id": "gpt-3.5-turbo-0301",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": undefined,
    "displayName": "gpt-3.5-turbo-0613",
    "enabled": false,
    "functionCall": false,
    "id": "gpt-3.5-turbo-0613",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": undefined,
    "displayName": "TTS-1",
    "enabled": false,
    "functionCall": false,
    "id": "tts-1",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": undefined,
    "displayName": "DALL·E 3",
    "enabled": false,
    "functionCall": false,
    "id": "dall-e-3",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": 16385,
    "displayName": "GPT-3.5 Turbo 1106",
    "enabled": false,
    "functionCall": true,
    "id": "gpt-3.5-turbo-1106",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": 128000,
    "displayName": "GPT-4 Turbo Preview 1106",
    "enabled": false,
    "functionCall": true,
    "id": "gpt-4-1106-preview",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": undefined,
    "displayName": "babbage-002",
    "enabled": false,
    "functionCall": false,
    "id": "babbage-002",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": undefined,
    "displayName": "tts-1-1106",
    "enabled": false,
    "functionCall": false,
    "id": "tts-1-1106",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": 128000,
    "displayName": "GPT 4 Turbo with Vision Preview",
    "enabled": false,
    "functionCall": false,
    "id": "gpt-4-vision-preview",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": true,
  },
  {
    "contextWindowTokens": 8192,
    "displayName": "Text Embedding 3 Small",
    "enabled": false,
    "functionCall": false,
    "id": "text-embedding-3-small",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": 128000,
    "displayName": "GPT 4 Turbo",
    "enabled": false,
    "functionCall": true,
    "id": "gpt-4",
    "maxOutput": 4096,
    "reasoning": false,
    "vision": true,
  },
  {
    "contextWindowTokens": undefined,
    "displayName": "text-embedding-ada-002",
    "enabled": false,
    "functionCall": false,
    "id": "text-embedding-ada-002",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": 16385,
    "displayName": "GPT-3.5 Turbo 0125",
    "enabled": false,
    "functionCall": true,
    "id": "gpt-3.5-turbo-0125",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
  {
    "contextWindowTokens": 8192,
    "displayName": "GPT-4 0613",
    "enabled": false,
    "functionCall": true,
    "id": "gpt-4-0613",
    "maxOutput": undefined,
    "reasoning": false,
    "vision": false,
  },
]
`;
